### PR TITLE
Add JSON parsing to Custom block

### PR DESCRIPTION
**Pull request by ammgws**
This builds on the work in #165:
- support for setting the icon has been added
- non-JSON output is passed through as-is which allows us to not break users' configs (block behaves the same as now).

This allows for a more versatile `custom` block and should solve the following issues such as #331, #414 and #454 as long as the user writes a wrapper script for their commands that outputs JSON.

### CODE REVIEW

**Comment by jcgruenhage** [comment]
This is a very neat solution, a lot better than mine.  Thanks!

**Comment by ammgws** [comment]
@jcgruenhage Sorry, I wanted to build off your commit but couldn't figure out how to cherry pick it.

Hopefully there's no blockers left for merging this.

**Comment by jcgruenhage** [comment]
No need to be sorry, my solution was a lot bulkier than yours anyway. @atheriel is there anything preventing this from being merged?

**Comment by jcgruenhage** [comment]
Just to make sure this isn't overlooked, I'll ping @greshake and @atheriel again. Merging this (or a review explaining why not) would be greatly appreciated :)

**Comment by ammgws** [comment]
Rebased and resolved conflicts. Also updated the README, but might have to polish it up a bit more.

**Comment by jcgruenhage** [comment]
Another month has passed since I pinged the maintainers, so...
image
@atheriel, @greshake 

This is a quite small contribution, this is not much to review. This feature has been implemented multiple times (see #454, #414, #166), people have wanted it for years. I know you've got other stuff to do, and I don't post this with the intention to make you feel bad. On the contrary, if I wasn't so happy with the rest of i3status-rust, I wouldn't bother. If you need some more hands on maintainance here, I can definitely help with testing and reviewing PRs.

**Comment by ammgws** [comment]
Probably not a good time given the whole COVID19 situation. Hopefully this can be merged soon though!

**Comment by omertuc** [comment]
If my understanding it correct, the code attempts to parse the output of the command as JSON, and if that fails, it fallbacks to the old behavior (old behavior being showing the raw output Info state, no icon, text)? 

Wouldn't it better to to have a configuration parameter for the "custom" block that specifies whether the command output is this special JSON format or raw text, while defaulting to raw text if unspecified?

This implicit-try-catch behavior is a little bit unintuitive IMO, and might be confusing for anyone not familiar enough with the code to understand. Especially in the current state where this behavior is not documented.

**Comment by jcgruenhage** [comment]
It is documented, the blocks.md file has been updated accordingly. I think adding a config option for this would be okay, but I think it's fine like this too

**Comment by jcgruenhage** [comment]
Another month has passed since I pinged the maintainers, so...
image
So, if you want me to just fuck-off, please tell me. @greshake, @atheriel: This project is really cool, I really like what you've done here, but the lack of any interaction here is quite annoying. So I'll just put it another way around: If I don't see a reply here for another month, I'll start looking into forking this.

**Comment by ammgws** [comment]
I just got push access today so I can move this forward.

**Comment by ammgws** [comment]
Ended up adding the config param as mentioned by @omertuc, and also fixed up the code (icon wasn't actually working).

File: blocks.md [hunk 1]
```diff
@@ -176,6 +177,11 @@ Key | Values | Required | Default
 
 Creates a block that display the output of custom shell commands.
 
+For further customisation, use the `json` option and have the shell command output valid JSON in the schema below:
+`{"icon": "ICON", state": "STATE", "text": "YOURTEXT"}`
```
Comment (omertuc) [inline blocks.md:181]: Missing "

Comment (ammgws) [reply blocks.md:181]: Cheers

**Pull request merged**